c/extras/dt_ftrl.cc

```diff
@@ -64,33 +64,28 @@ void Ftrl::fit(const DataTable* dt_X, const DataTable* dt_y) {
   auto c_y = static_cast<BoolColumn*>(dt_y->columns[0]);
   auto d_y = c_y->elements_r();
 
-  // Do training for `nepochs`.
-  for (size_t i = 0; i < params.nepochs; ++i) {
-    double total_loss = 0;
+  const RowIndex ri_X = dt_X->rowindex;
+  const RowIndex ri_y = c_y->rowindex();
 
+  // Do training for `nepochs`.
+  for (size_t e = 0; e < params.nepochs; ++e) {
     #pragma omp parallel num_threads(config::nthreads)
     {
       // Array to store hashed column values and their interactions.
       uint64ptr x = uint64ptr(new uint64_t[nfeatures]);
       size_t ith = static_cast<size_t>(omp_get_thread_num());
       size_t nth = static_cast<size_t>(omp_get_num_threads());
 
-      for (size_t j = ith; j < dt_X->nrows; j += nth) {
-        if (ISNA<int8_t>(d_y[j])) continue;
-        bool y = d_y[j];
-        hash_row(x, j);
-        double p = predict_row(x);
-        update(x, p, y);
-
-        double loss = logloss(p, y);
-        #pragma omp atomic update
-        total_loss += loss;
-        if ((j+1) % REPORT_FREQUENCY == 0) {
-          printf("Training epoch: %zu\tRow: %zu\tPrediction: %f\t"
-                 "Current loss: %f\tAverage loss: %f\n",
-                 i, j+1, p, loss, total_loss / (j+1));
+      ri_X.iterate(ith, dt_X->nrows, nth,
+        [&](size_t i, size_t j) {
```

Comment:
When iterating over a RowIndex you should also be prepared to encounter `j == RowIndex::NA` indicating that the value in that column is NA.

Reply:
`i`th row of the result is the `j`th row of the source. Indices `i` are always consecutive (at least in single-threaded mode): 0, 1, 2, 3, ... Indices `j` can be anything, including `-1` (which is NA) indicating that there is no source row for that `i`th place.

This can happen for example when (1) the number of rows of a frame is increased; (2) several frames with different number of rows are force-cbound into a single frame; (3) an outer join is performed but some rows are not matched in either of the joined frames.